File: trunk/proj/Android/src/com/jasonzqshen/familyaccounting/core/masterdata/BankKeyMasterDataFactory.java
package com.jasonzqshen.familyaccounting.core.masterdata;

import org.w3c.dom.Element;

import com.jasonzqshen.familyaccounting.core.CoreDriver;
import com.jasonzqshen.familyaccounting.core.exception.IdentityInvalidChar;
import com.jasonzqshen.familyaccounting.core.exception.IdentityNoData;
import com.jasonzqshen.familyaccounting.core.exception.IdentityTooLong;
import com.jasonzqshen.familyaccounting.core.exception.MasterDataFileFormatException;
import com.jasonzqshen.familyaccounting.core.exception.MasterDataIdentityExists;
import com.jasonzqshen.familyaccounting.core.exception.NullValueNotAcceptable;
import com.jasonzqshen.familyaccounting.core.exception.ParametersException;
import com.jasonzqshen.familyaccounting.core.exception.runtime.SystemException;
import com.jasonzqshen.familyaccounting.core.utils.CoreMessage;
import com.jasonzqshen.familyaccounting.core.utils.MessageType;

public class BankKeyMasterDataFactory extends MasterDataFactoryBase {

	public BankKeyMasterDataFactory(CoreDriver coreDriver) {
		super(coreDriver);
	}

	@Override
	public MasterDataBase createNewMasterDataBase(MasterDataIdentity id,
			String descp, Object... objects) throws ParametersException,
			MasterDataIdentityExists {
		if (objects.length > 0) {
			throw new ParametersException(String.format(
					CoreMessage.ERR_PARAMETER_LENGTH, 0, objects.length));
		}

		// check identity is duplicated
		if (_list.containsKey(id)) {
			throw new MasterDataIdentityExists();
		}

		BankKeyMasterData bankKey;
		try {
			bankKey = new BankKeyMasterData(_coreDriver, id, descp);
		} catch (NullValueNotAcceptable e) {
			throw new SystemException(e);
		}

		this._containDirtyData = true;
		this._list.put(id, bankKey);

		_coreDriver.logDebugInfo(this.getClass(), 47,
				String.format("Create bank key (%s).", bankKey.toXML()),
				MessageType.INFO);
		return bankKey;
	}

	@Override
	public MasterDataBase parseMasterData(CoreDriver coreDriver, Element elem)
			throws MasterDataFileFormatException {
		String id = elem.getAttribute(MasterDataUtils.XML_ID);
		String descp = elem.getAttribute(MasterDataUtils.XML_DESCP);

		MasterDataIdentity identity;
		try {
			identity = new MasterDataIdentity(id.toCharArray());
			BankKeyMasterData bankKey = (BankKeyMasterData) this
					.createNewMasterDataBase(identity, descp);

			_coreDriver.logDebugInfo(this.getClass(), 61,
					String.format("Parse bank key (%s).", bankKey.toXML()),
					MessageType.INFO);
			return bankKey;
		} catch (IdentityTooLong e) {
			_coreDriver.logDebugInfo(this.getClass(), 150,
					"Master data identity is too long.", MessageType.ERROR);
			throw new MasterDataFileFormatException(MasterDataType.BANK_KEY);
		} catch (IdentityNoData e) {
			_coreDriver
					.logDebugInfo(this.getClass(), 154,
							"Master data identity is with no value.",
							MessageType.ERROR);
			throw new MasterDataFileFormatException(MasterDataType.BANK_KEY);
		} catch (IdentityInvalidChar e) {
			_coreDriver.logDebugInfo(this.getClass(), 160,
					"Invalid character in identity.", MessageType.ERROR);
			throw new MasterDataFileFormatException(MasterDataType.BANK_KEY);
		} catch (ParametersException e) {
			_coreDriver.logDebugInfo(this.getClass(), 164,
					"Function parameter set error: " + e.toString(),
					MessageType.ERROR);
			throw new SystemException(e);
		} catch (MasterDataIdentityExists e) {
			_coreDriver.logDebugInfo(this.getClass(), 168,
					"Master data identity duplicated.", MessageType.ERROR);
			throw new MasterDataFileFormatException(MasterDataType.BANK_KEY);
		}

	}

}

File: trunk/proj/WinStore_FamilyAccounting/WinStore_FamilyAccountingCore/MasterData/BankKeyMasterDataFactory.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Xml.Linq;
using WinStore_FamilyAccountingCore.Exceptions;
using WinStore_FamilyAccountingCore.Exceptions.FormatExceptions;
using WinStore_FamilyAccountingCore.Utilities;

namespace WinStore_FamilyAccountingCore.MasterData
{
    public class BankKeyMasterDataFactory : MasterDataFactoryBase
    {
        /// <summary>
        /// 
        /// </summary>
        /// <param name="coreDriver"></param>
        /// <param name="management"></param>
        /// <exception cref="ArgumentNullException">Argument is null</exception>
        public BankKeyMasterDataFactory(CoreDriver coreDriver,
            MasterDataManagement management)
            : base(coreDriver, management)
        {
        }

        /// <summary>
        /// Create new master data 
        /// </summary>
        /// <param name="identity"></param>
        /// <param name="descp"></param>
        /// <param name="?"></param>
        /// <returns></returns>
        /// <exception cref="ParametersException">Parameters Exception</exception>
        /// <exception cref="MasterDataIdentityExists">Duplicated master data identity exists </exception>
        /// <exception cref="MasterDataIdentityNotDefined">Master data identity is not defined</exception>
        /// <exception cref="SystemException">Bug</exception>
        public override MasterDataBase CreateNewMasterDataBase(MasterDataIdentity id
            , string descp, params object[] objects)
        {
            if (id == null
                || descp == null)
            {
                throw new ArgumentNullException();
            }
            if (objects.Length > 0)
            {
                throw new ParametersException(0, objects.Length);
            }

            // check identity is duplicated
            if (_list.ContainsKey(id))
            {
                throw new MasterDataIdentityExists();
            }

            BankKeyMasterData bankKey = new BankKeyMasterData(_coreDriver, _management, id, descp);
            this._containDirtyData = true;
            this._list.Add(id, bankKey);

            // raise create master data
            _coreDriver.ListenerMgmt.CreateMasterData(this, bankKey);
            _coreDriver.logDebugInfo(this.GetType(), 47,
                    String.Format("Create bank key ({0}).", bankKey.Identity.ToString()),
                    MessageType.INFO);
            return bankKey;
        }

        /// <summary>
        /// Parse master data from XML
        /// </summary>
        /// <param name="coreDriver"></param>
        /// <param name="elem"></param>
        /// <returns></returns>
        /// <exception cref="MasterDataFileFormatException">Master Data file exception</exception>
        /// <exception cref="ArgumentNullException">Argument is null</exception>
        /// <exception cref="SystemException">Bug</exception>
        public override MasterDataBase ParseMasterData(CoreDriver coreDriver, XElement elem)
        {
            XAttribute idAttr = elem.Attribute(MasterDataUtils.XML_ID);
            XAttribute descpAttr = elem.Attribute(MasterDataUtils.XML_DESCP);

            MasterDataIdentity identity;
            try
            {
                identity = new MasterDataIdentity(idAttr.Value);
                BankKeyMasterData bankKey = (BankKeyMasterData)this
                        .CreateNewMasterDataBase(identity, descpAttr.Value);

                _coreDriver.logDebugInfo(this.GetType(), 61,
                        String.Format("Parse bank key ({0}).", bankKey.Identity.ToString()),
                        MessageType.INFO);
                return bankKey;
            }
            catch (IdentityTooLong e)
            {
                _coreDriver.logDebugInfo(this.GetType(), 150, e.Message, MessageType.ERRO);
                throw new MasterDataFileFormatException(MasterDataType.BANK_KEY);
            }
            catch (IdentityNoData e)
            {
                _coreDriver
                        .logDebugInfo(this.GetType(), 154,
                                e.Message,
                                MessageType.ERRO);
                throw new MasterDataFileFormatException(MasterDataType.BANK_KEY);
            }
            catch (IdentityInvalidChar e)
            {
                _coreDriver.logDebugInfo(this.GetType(), 160, e.Message, MessageType.ERRO);
                throw new MasterDataFileFormatException(MasterDataType.BANK_KEY);
            }
            catch (ParametersException e)
            {
                _coreDriver.logDebugInfo(this.GetType(), 164,
                        "Function parameter set error: " + e.Message,
                        MessageType.ERRO);
                throw new SystemException(e);
            }
            catch (MasterDataIdentityExists e)
            {
                _coreDriver.logDebugInfo(this.GetType(), 168, e.Message, MessageType.ERRO);
                throw new MasterDataFileFormatException(MasterDataType.BANK_KEY);
            }
        }
    }
}
